### PR TITLE
Tag retention webhook support

### PR DESCRIPTION
**Pull request by tedgxt**
Signed-off-by: guanxiatao <guanxiatao@corp.netease.com>

### CODE REVIEW

**Comment by Purushotham233** [comment]
@tedgxt, could you add the reference of the issue this PR is fixing?

File: src/controller/event/handler/webhook/artifact/replication.go [hunk 1]
```diff
@@ -8,14 +8,14 @@ import (
 	commonModels "github.com/goharbor/harbor/src/common/models"
 	"github.com/goharbor/harbor/src/controller/event"
 	"github.com/goharbor/harbor/src/controller/event/handler/util"
+	ctlModel "github.com/goharbor/harbor/src/controller/event/model"
```
Comment (wy65701436) [inline src/controller/event/handler/webhook/artifact/replication.go:11]: evtModel?

File: src/controller/event/handler/webhook/artifact/retention.go [hunk 2]
```diff
@@ -0,0 +1,148 @@
...
+// Handle ...
+func (r *RetentionHandler) Handle(value interface{}) error {
+	if !config.NotificationEnable() {
+		log.Debug("notification feature is not enabled")
+		return nil
+	}
+
+	trEvent, ok := value.(*event.RetentionEvent)
+	if !ok {
+		return errors.New("invalid tag retention event type")
```
Comment (wy65701436) [inline src/controller/event/handler/webhook/artifact/retention.go:31]: use the `lib/errors` instead

File: src/controller/event/handler/webhook/artifact/retention.go [hunk 3]
```diff
@@ -0,0 +1,148 @@
...
+	}
+	if trEvent == nil {
+		return errors.New("nil tag retention event")
+	}
+	if len(trEvent.Deleted) == 0 {
```
Comment (wy65701436) [inline src/controller/event/handler/webhook/artifact/retention.go:36]: zero deletion is actually a event, why drop it?

File: src/controller/event/handler/webhook/artifact/retention.go [hunk 4]
```diff
@@ -0,0 +1,148 @@
...
+			},
+		},
+	}
+
+	for _, v := range event.Deleted {
+		target := v.Target
+		deletedArtifact := &ctlModel.ArtifactInfo{
+			Type:       target.Kind,
+			Status:     event.Status,
+			NameAndTag: target.Repository + ":" + target.Tags[0],
```
Comment (wy65701436) [inline src/controller/event/handler/webhook/artifact/retention.go:132]: be careful the `panic`(runtime error: index out of range) here.

File: src/core/service/notifications/jobs/handler.go [hunk 5]
```diff
@@ -205,6 +206,23 @@ func (h *Handler) HandleRetentionTask() {
 			h.SendInternalServerError(err)
 			return
 		}
+
+		e := &event.Event{}
+		metaData := &metadata.RetentionMetaData{
+			Total:    retainObj.Total,
+			Retained: retainObj.Retained,
+			Deleted:  retainObj.Deleted,
+			Status:   "SUCCESS",
```
Comment (wy65701436) [inline src/core/service/notifications/jobs/handler.go:215]: always success?

File: src/controller/event/handler/webhook/artifact/replication.go (hunk 1 above)
Comment (tedgxt) [reply src/controller/event/handler/webhook/artifact/replication.go:11]: fixed

File: src/controller/event/handler/webhook/artifact/retention.go (hunk 2 above)
Comment (tedgxt) [reply src/controller/event/handler/webhook/artifact/retention.go:31]: fixed

Comment (tedgxt) [reply src/controller/event/handler/webhook/artifact/retention.go:31]: fixed

File: src/controller/event/handler/webhook/artifact/retention.go (hunk 3 above)
Comment (tedgxt) [reply src/controller/event/handler/webhook/artifact/retention.go:36]: > zero deletion is actually a event, why drop it?

project info can only be found in `Deleted` ， if there is no deletion，we cannot get the webhook policy of the project

File: src/controller/event/handler/webhook/artifact/retention.go (hunk 4 above)
Comment (tedgxt) [reply src/controller/event/handler/webhook/artifact/retention.go:132]: > be careful the `panic`(runtime error: index out of range) here.

fixed

File: src/core/service/notifications/jobs/handler.go (hunk 5 above)
Comment (tedgxt) [reply src/core/service/notifications/jobs/handler.go:215]: > always success?

Tag rentention info will be received only when task begin to  run.      when the task status with success received, there is no retention info with it.

**Pull request merged**